File: Second_Year/DAAA(Design_and_Analysis_of_Algorithms)/Assignment_1.cpp
#include <iostream>
#include <chrono>
using namespace std;
using namespace std::chrono;

int main() {
    int n;

    cout << "Enter number of elements: ";
    cin >> n;

    if (n <= 0) {
        cout << "Array is empty!" << endl;
        return 0;
    }

    int arr[n];

    cout << "Enter elements:\n";
    for(int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    auto start = high_resolution_clock::now();

    int min = arr[0];
    int max = arr[0];

    for(int i = 1; i < n; i++) {
        if(arr[i] < min)
            min = arr[i];
        if(arr[i] > max)
            max = arr[i];
    }

    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<nanoseconds>(stop - start);

    cout << "\nMinimum element: " << min << endl;
    cout << "Maximum element: " << max << endl;

    cout << "Execution time: " << duration.count() << " nanoseconds" << endl;

    long long storage = sizeof(arr) + sizeof(min) + sizeof(max) + sizeof(n);
    cout << "Storage used: " << storage << " bytes" << endl;

    return 0;
}

File: Second_Year/DAAA(Design_and_Analysis_of_Algorithms)/Assignment_1.java
import java.util.ArrayList;
import java.util.Scanner;

public class Assignment_1 {

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        ArrayList<Integer> arr = new ArrayList<>();

        System.out.print("Enter number of elements: ");
        int n = sc.nextInt();

        if (n <= 0) {
            System.out.println("Array is empty!");
            return;
        }

        System.out.println("Enter elements:");
        for (int i = 0; i < n; i++) {
            arr.add(sc.nextInt());
        }

        long startTime = System.nanoTime();

        int min = arr.get(0);
        int max = arr.get(0);

        for (int i = 1; i < n; i++) {
            if (arr.get(i) < min)
                min = arr.get(i);
            if (arr.get(i) > max)
                max = arr.get(i);
        }

        long endTime = System.nanoTime();
        long duration = endTime - startTime;

        System.out.println("\nMinimum element: " + min);
        System.out.println("Maximum element: " + max);

        System.out.println("Execution time: " + duration + " nanoseconds");

        // Approximate storage calculation
        int storage = (n * Integer.BYTES) + Integer.BYTES * 3;
        System.out.println("Approximate storage used: " + storage + " bytes");

        sc.close();
    }
}
